storage/src/main/java/com/epam/aidial/core/storage/service/ResourceService.java

```diff
@@ -310,16 +311,25 @@ private Pair<ResourceItemMetadata, String> getResourceWithMetadata(ResourceDescr
 
     @Nullable
     public String getResource(ResourceDescriptor descriptor) {
-        return getResource(descriptor, true);
+        return getResource(descriptor, EtagHeader.ANY, true);
     }
 
     @Nullable
-    private String getResource(ResourceDescriptor descriptor, boolean lock) {
-        Pair<ResourceItemMetadata, String> result = getResourceWithMetadata(descriptor, lock);
+    private String getResource(ResourceDescriptor descriptor, EtagHeader etag, boolean lock) {
+        Pair<ResourceItemMetadata, String> result = getResourceWithMetadata(descriptor, etag, lock);
         return (result == null) ? null : result.getRight();
     }
 
-    public ResourceStream getResourceStream(ResourceDescriptor resource) throws IOException {
+    public ResourceStream getResourceStream(ResourceDescriptor resource, EtagHeader etagHeader) throws IOException {
+        ResourceStream resourceStream = getResourceStream(resource);
+        if (resourceStream != null) {
+            etagHeader.validate(resourceStream.etag);
```

Comment:
Rework please. It can throw an exception. You need to close the stream in this case and rethrow.

Reply:
fixed